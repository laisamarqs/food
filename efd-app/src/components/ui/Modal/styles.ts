import styled from 'styled-components'
import { colors, breakpoints } from '../../../styles'
import { ButtonContainer } from '../Button/styles'

export const Modal = styled.div`
  position: fixed;
  top: 0;
  left: 0;
  z-index: 1;
  width: 100%;
  height: 100%;
  display: flex;
  align-items: center;
  justify-content: center;
`

export const ModalContainer = styled.div`
  background-color: ${colors.primaryDark};
  width: 64rem;
  height: 21.5rem;
  padding: 2rem;
  display: flex;
  position: relative;
  z-index: 1;

  img {
    width: 17.5rem;
    height: 17.5rem;
    object-fit: cover;
  }

  @media only screen and (max-width: ${breakpoints.desktop}) {
    width: 70%;

    img {
      height: 250px;
    }
  }

  @media only screen and (max-width: ${breakpoints.tablet}) {
    width: 70%;
    height: auto;
    display: grid;
    padding: 1.8rem;

    img {
      width: 100%;
      margin-top: 1rem;
    }
  }

  @media only screen and (max-width: ${breakpoints.mobile}) {
    width: 80%;
    padding: 1.2rem;
    padding-top: 2rem;

    img {
      width: 100%;
      height: 165px;
    }
  }
`
export const Close = styled.button`
  border: none;
  background: transparent;
  top: 0.5rem;
  right: 0.5rem;
  position: absolute;
  cursor: pointer;
  width: 48px;
  height: 48px;

  img {
    width: 1rem;
    height: 1rem;
  }

  @media only screen and (max-width: ${breakpoints.tablet}) {
    top: 0;
    right: 0;
  }
`

export const Infos = styled.div`
  color: #fff;
  margin-left: 1.4rem;
  display: grid;

  h3 {
    font-size: 1.125rem;
    font-weight: 900;
  }

  p {
    font-size: 0.875rem;
    font-weight: 400;
    line-height: 1.375rem;
    padding-bottom: 1rem;
  }

  ${ButtonContainer} {
    width: 14.625rem;
    padding: 0;
  }

  @media only screen and (max-width: ${breakpoints.tablet}) {
    margin: 0;
    h3 {
      padding: 0.5rem 0;
    }

    ${ButtonContainer} {
      width: 100%;
      margin-top: 1rem;
      padding: 0.5rem;
    }
  }
`