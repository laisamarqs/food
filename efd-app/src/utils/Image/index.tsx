import { removeLoading } from '..'

export type ImageProps = {
  alt: string | undefined
  src: string | undefined
  width?: number
  height?: number
}

const Image = ({
  src,
  alt,
  width = 100,
  height = 100,
  ...props
}: ImageProps) => (
  <img
    loading="lazy"
    src={src}
    alt={alt}
    onLoad={removeLoading}
    className="loading"
    width={width}
    height={height}
    {...props}
  />
)

export default Image