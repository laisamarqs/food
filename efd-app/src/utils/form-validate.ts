import { FormikProps } from 'formik'

type Field =
  | 'address'
  | 'cardCode'
  | 'cardName'
  | 'cardNumber'
  | 'city'
  | 'expiresMonth'
  | 'expiresYear'
  | 'name'
  | 'number'
  | 'zipCode'

  const getFieldStatus = (field: Field, form: FormikProps<FormValues>) => {
    const isTouched = field in form.touched
    const isInvalid = field in form.errors
  
    return isTouched && isInvalid
  }
  
  export const getErrorMessage = (
    field: Field,
    form: FormikProps<FormValues>
  ): string => {
    const hasError = getFieldStatus(field, form)
  
    const errorMessage = form.errors[field]
  
    if (hasError && errorMessage) return errorMessage
    return ''
  }