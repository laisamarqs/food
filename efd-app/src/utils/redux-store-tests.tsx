import { PreloadedState } from '@reduxjs/toolkit'
import { RenderOptions, render } from '@testing-library/react'
import { PropsWithChildren, ReactElement } from 'react'
import { Provider } from 'react-redux'

import { RootState, AppStore, configStore } from '../store'
interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export function renderWithProvider(
  element: ReactElement,
  {
    preloadedState = {},
    store = configStore(preloadedState),
    ...moreOptions
  }: ExtendedRenderOptions = {}
) {
  function Wrapper({ children }: PropsWithChildren<{}>): JSX.Element {
    return <Provider store={store}>{children}</Provider>
  }

  return {
    store,
    ...render(element, {
      wrapper: Wrapper,
      ...moreOptions
    })
  }
}