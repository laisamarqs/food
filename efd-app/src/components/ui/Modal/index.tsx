import { useDispatch, useSelector } from 'react-redux'
import { Overlay } from '../../../styles'
import { formatPrice } from '../../../utils'
import Button from '../Button'
import * as Style from './styles'
import { closeModal } from '../../../store/reducers/modal'
import closeIcon from '../../../assets/images/close.png'
import Image from '../../../utils/Image'
import { openSideBar } from '../../../store/reducers/sideBar'
import { addToCart } from '../../../store/reducers/cart'
import { RootState } from '../../../store'

const Modal = () => {
  const dispatch = useDispatch()
  const { food } = useSelector((state: RootState) => state.modal)

  const addFoodToCart = () => {
    if (food) {
      dispatch(addToCart(food))
      dispatch(closeModal())
      dispatch(openSideBar())
    } else {
      alert('Algo deu errado, tente novamente...')
    }
  }

  if (!food) return <span>Parece que não há um produto aqui :(</span>

  return (
    <Style.Modal>
      <Style.ModalContainer>
        <Image src={food.foto} alt={food.nome} />
        <Style.Infos>
          <h3>{food.nome}</h3>
          <p>{food.descricao}</p>
          <span>Serve: {food.porcao}</span>
          <Button
            title="Adicionar ao carrinho"
            background="secundary"
            type="button"
            onClick={addFoodToCart}
            label={`Adicionar ao carrinho - ${formatPrice(food?.preco)}`}
          />
        </Style.Infos>
        <Style.Close onClick={() => dispatch(closeModal())}>
          <img src={closeIcon} alt="Icone de fechar" />
        </Style.Close>
      </Style.ModalContainer>
      <Overlay data-testid="overlay" onClick={() => dispatch(closeModal())} />
    </Style.Modal>
  )
}

export default Modal