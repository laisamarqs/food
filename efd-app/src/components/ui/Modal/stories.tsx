import type { Meta } from '@storybook/react'

import Modal from '.'

import mockStore from '../../../mocks/redux-store'
import { Provider } from 'react-redux'

const meta = {
  title: 'ui/Modal',
  component: Modal,
  parameters: {
    layout: 'centered'
  }
} satisfies Meta<typeof Modal>

// eslint-disable-next-line react-refresh/only-export-components
export default meta

export const Template = () => {
  const store = mockStore({
    modal: {
      food: {
        foto: 'https://upload.wikimedia.org/wikipedia/commons/1/1e/Comida_de_mercado.jpg',
        preco: 98.9,
        id: 1,
        nome: 'Comida A',
        descricao:
          'A La Dolce Vita Trattoria leva a autêntica cozinha italiana até você! Desfrute de massas caseiras, pizzas deliciosas e risotos incríveis, tudo no conforto do seu lar. Entrega rápida, pratos bem embalados e sabor inesquecível. Peça já!',
        porcao: 'de 2 a 3 pessoas'
      }
    }
  })
  return (
    <Provider store={store}>
      <Modal />
    </Provider>
  )
}