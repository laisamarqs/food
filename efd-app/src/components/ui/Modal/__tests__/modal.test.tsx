import '@testing-library/jest-dom'
import { fireEvent, screen } from '@testing-library/react'
import { renderWithProvider } from '../../../../utils/redux-store-tests'
import Modal from '..'
import { expect } from 'vitest'

const renderComponent = () => {
  const { store, container } = renderWithProvider(<Modal />, {
    preloadedState: {
      modal: {
        food: {
          foto: '/img.png',
          preco: 98.9,
          id: 1,
          nome: 'Comida A',
          descricao: 'descrição comida A',
          porcao: '3 pessoas'
        },
        isModalOpen: true
      }
    }
  })

  return { store, container }
}

describe('<Modal />', () => {
  it('should render correctly when state is true', () => {
    const { container } = renderComponent()

    expect(
      screen.getByRole('heading', { name: 'Comida A' })
    ).toBeInTheDocument()

    expect(container.firstChild).toMatchSnapshot()
  })
  it('should render a paragraph when there is no food in the state', () => {
    renderWithProvider(<Modal />)

    expect(
      screen.getByText('Parece que não há um produto aqui :(')
    ).toBeInTheDocument()
  })

  it('should call the addFoodToCart function when clicking the add to cart button', () => {
    const { store } = renderComponent()

    const addFoodButton = screen.getByRole('button', {
      name: /Adicionar ao carrinho/i
    })

    fireEvent.click(addFoodButton)

    expect(store.getState().cart.itemsCart[0].nome).toBe('Comida A')
    expect(store.getState().modal.isModalOpen).toBeFalsy()
    expect(store.getState().sideBar.sideBarIsOpen).toBeTruthy()
  })

  it('should close modal when pressing the close button', () => {
    const { store } = renderComponent()

    const button = screen.getByRole('button', { name: 'Icone de fechar' })

    fireEvent.click(button)

    expect(store.getState().modal.isModalOpen).toBeFalsy()
  })

  it('should close modal when pressing the overlay', () => {
    const { store } = renderComponent()

    const overlayContainer = screen.getByTestId('overlay')

    fireEvent.click(overlayContainer)

    expect(store.getState().modal.isModalOpen).toBeFalsy()
  })
})