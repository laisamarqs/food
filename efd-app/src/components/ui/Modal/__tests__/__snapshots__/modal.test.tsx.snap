// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`<Modal /> > should render correctly when state is true 1`] = `
<div
  class="sc-iGgWBj oFAYE"
>
  <div
    class="sc-gsFSXq cVMKeN"
  >
    <img
      alt="Comida A"
      class="loading"
      height="100"
      loading="lazy"
      src="/img.png"
      width="100"
    />
    <div
      class="sc-imWYAI hLZLGI"
    >
      <h3>
        Comida A
      </h3>
      <p>
        descrição comida A
      </p>
      <span>
        Serve: 
        3 pessoas
      </span>
      <button
        class="sc-dcJsrY bDWePn"
        title="Adicionar ao carrinho"
        type="button"
      >
        Adicionar ao carrinho - R$ 98,90
      </button>
    </div>
    <button
      class="sc-kAyceB dCjlVK"
    >
      <img
        alt="Icone de fechar"
        src="/src/assets/images/close.png"
      />
    </button>
  </div>
  <div
    class="sc-gEvEer cSvWod"
    data-testid="overlay"
  />
</div>
`;